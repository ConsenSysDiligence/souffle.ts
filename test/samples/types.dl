// A,B,Magic,Name,Translate
// https://souffle-lang.github.io/types#float-type

.decl Name(n: symbol)
Name("Hans").
Name("Gretl").

.decl Translate(n: symbol , o: number)
// .output Translate
Translate(x,ord(x)) :- Name(x).

.decl Magic(x:number, y:unsigned)
Magic(-1,1).
// .output Magic

// https://souffle-lang.github.io/types#equivalence-types

.type myNumber = number

// https://souffle-lang.github.io/types#subtypes

.type myEvenNumber <: number
.type myMultiplesOfFour <: myEvenNumber

// https://souffle-lang.github.io/types#union-type

.type City <: symbol
.type Town <: symbol
.type Village <: symbol
.type Place = City | Town | Village

// https://souffle-lang.github.io/types#record-types

.type List = [
    head : number,
    tail : List
]
.decl B(x : List)
B(nil).
B([1,nil]).
B([2,[3,nil]]).
// .output B

// https://souffle-lang.github.io/types#algebraic-data-types-adt

.type Expression = Number { x : number }
                | Variable { v : symbol}
                | Add {e_1 : Expression, e_2 :Expression}
                | Imaginary {}

.decl A(x:Expression)

A($Number(10)).
A($Add($Number(10),$Imaginary())).
A($Add($Number(10), $Variable("x"))).
A($Number(x+1)) :- A($Number(x)), x < 20.

// .output A
