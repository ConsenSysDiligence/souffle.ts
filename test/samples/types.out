/// A
===============
x
$Number(10)
$Add($Number(10), $Imaginary())
$Add($Number(10), $Variable(x))
$Number(11)
$Number(12)
$Number(13)
$Number(14)
$Number(15)
$Number(16)
$Number(17)
$Number(18)
$Number(19)
$Number(20)
/// B
===============
x
nil
[1, nil]
[2, [3, nil]]
/// Magic
===============
x y z
-1    1    2.71799994
/// Name
===============
n
Hans
Gretl
/// Translate
===============
n o
Hans    1
Gretl    2